src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp

```diff
@@ -0,0 +1,227 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <type_traits>
+
+#include "DataStructures/ApplyMatrices.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/SliceIterator.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Domain/Structure/Direction.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeArray.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace evolution::dg {
+namespace detail {
+template <typename TagsList>
+struct NumberOfIndependentComponents;
+
+template <typename... Tags>
+struct NumberOfIndependentComponents<tmpl::list<Tags...>> {
+  static constexpr size_t value = (... + Tags::type::size());
+};
+}  // namespace detail
+
+/*!
+ * \brief Projects a `Variables` of volume data to a contiguous subset of
+ * a boundary `Variables`
+ *
+ * The `volume_fields` are all projected into the `face_fields` in the direction
+ * `direction`. The variables in the volume (`VolumeVarsTagsList`) must be a
+ * subset of the list of the tags on the face (`FaceVarsTagsList`). Not just the
+ * individual variables, but the entire list must be a subset. This is because
+ * the projection is applied on all of the tensor components of the volume
+ * variables and is written into contiguous memory on the boundary.
+ *
+ * In general, this function will be used for projecting all the evolved
+ * variables or all the volume fluxes to the faces. The function
+ * `evolution::dg::project_tensors_to_boundary()` should be used for projecting
+ * individual tensors to the face.
+ *
+ * \note This function works for both Gauss and Gauss-Lobatto uniform meshes.
+ */
+template <typename VolumeVarsTagsList, typename FaceVarsTagsList, size_t Dim>
+void project_contiguous_data_to_boundary(
+    const gsl::not_null<Variables<FaceVarsTagsList>*> face_fields,
+    const Variables<VolumeVarsTagsList>& volume_fields,
+    const Mesh<Dim>& volume_mesh, const Direction<Dim>& direction) noexcept {
+  static_assert(tmpl::size<VolumeVarsTagsList>::value != 0,
+                "Must have non-zero number of volume fields");
+  static_assert(tmpl::size<FaceVarsTagsList>::value != 0,
```

Comment:
- Instead of checking `!= 0`, could you check `>= tmpl::size<VolumeVarsTagsList>::value`? This would fail faster for lists of face tags that are obviously not supersets of the volume tags...
- In the same line of reasoning... would it make sense to add static_asserts that `tmpl::back<VolumeVarsTagsList>` and `tmpl::front<VolumeVarsTagsList>` are both in `FaceVarsTagsList`? Maybe this is getting to overkill.